Add new updated limit method that takes word boundries into consideration

itsmewes opened the pull request:
Currently for limiting string lengths you can use `Str::limit` and `Str::words` which are great options. The problem that I ran into is that I would like to limit the string to whole words. `Str::limit` can cut into a word and `Str::words` can be tough to use since the length of a string  is wildly different depending on word sizes. This new `Str::limitTillWord` takes in characters like `Str::limit` but will make sure that it doesn't cut into any new words but completes on a word boundary. I believe this makes string lengths a lot more consistent when working with word boundaries. The downside to this is that there is no `Multibyte String (mb_*)` string version of `wordwrap`.

wesleymartincc committed:
Add new updated limit method that takes word boundries into consideration

wesleymartincc committed:
Add new updated limit method that takes word boundries into consideration

wesleymartincc committed:
Rather use regex to remove any non word characters

wesleymartincc committed:
Rather use regex to remove any non word characters

ahinkle commented:
Seems a little niche. Might be better in userland. 👍

henzeb commented:
Sounds interesting. The name is a bit lengthy ...

itsmewes commented:
> Seems a little niche. Might be better in userland. 👍

That's a fair point. Personally I prefer having the limit always be on a word boundry but that is personal preference.

itsmewes commented:
> Sounds interesting. The name is a bit lengthy ...

Yeah, naming is hard. Open to new names 🙂

antonkomarev commented:
May it be an extra enum param in limit method? Default to limit as it works now, another value to trim to words. But maybe it's better to change signature in 12.x

rodrigopedra commented:
`Str::wordLimit()`?

Really like it, I use a much worse implementation in many projects 😅 (thus the naming suggestion)

So, even if this doesn't make the cut, thank you so much for sharing

Specially for sending error reports through e-mail, or logging errors into constrained outputs, limiting by words is more helpful

itsmewes commented on src/Illuminate/Support/Str.php:
```diff
@@ -636,6 +636,27 @@ public static function limit($value, $limit = 100, $end = '...')
+     */
+    public static function limitTillWord($value, $limit = 100, $end = '...')
+    {
+        if (mb_strwidth($value, 'UTF-8') <= $limit) {
+            return $value;
+        }
+
+        $str = trim(preg_replace('/[\n\r]+/', ' ', strip_tags($value)));
+        $wordWrap = wordwrap($str, $limit);
+        $sentences = explode("\n", $wordWrap);
+
+        return preg_replace('/\W$/', '', $sentences[0]) . $end;
```
This is a proposed alternative to this as well which I believe would handle multibyte strings.

```suggestion
        if (mb_strwidth($value, 'UTF-8') <= $limit) {
            return $value;
        }

        $value = trim(preg_replace('/[\n\r]+/', ' ', strip_tags($value)));
        $trimmed = rtrim(mb_strimwidth($value, 0, $limit, '', 'UTF-8'));

        if (mb_substr($value, $limit, 1, 'UTF-8') === ' ') {
            return $trimmed.$end;
        }

        return preg_replace("/(.*)\s.*/", '$1', $trimmed).$end;
```

itsmewes commented:
> May it be an extra enum param in limit method? Default to limit as it works now, another value to trim to words. But maybe it's better to change signature in 12.x
> 
> Something like `$mode`

This was actually my first thought, but then decided to not touch what was already there. I do think having some form of flag on the limit method would be first prize.

crynobone commented:
Hi @itsmewes, can you rebase this PR to latest upstream

itsmewes committed:
Merge branch 'laravel:11.x' into str/respect-word-limit

crynobone committed:
[11.x] Test Improvements (#52122)

fixes tests from
https://github.com/laravel/framework/commit/f5f9383af919ef74b00eacb23b4a88fcf390743c
commit

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

bakerkretzmar committed:
[11.x] Fix SQLite schema dumps containing internal `sqlite_*` objects (#52135)

* Add failing test

* Remove all `CREATE TABLE sqlite_*` statements from schema dump

* Wip

TENIOS committed:
Fix typo in `<!DOCTYPE>` declaration (#52134)

calebdw committed:
fix: pluck generics (#52109)

lewislarsen committed:
[11.x] Include 'success' console component (#52112)

* add success console component

* Update Line.php

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

hafezdivandari committed:
[11.x] Fix dumping migrations table with schema or prefixed name (#52098)

* fix dump with schema and prefixed migrations table

* formatting

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

jasonmccreary committed:
Add `assertSentTo` shorthand (#52083)

* Add `assertSentTo` shorthand

* Add multiple example to test

* Overload existing method

* Add negative assertions

taylorotwell committed:
Update facade docblocks

taylorotwell committed:
wip

taylorotwell committed:
Update facade docblocks

crynobone committed:
[11.x] Test Improvements (#52139)

* [11.x] Test Improvements

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

---------

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

kevinb1989 committed:
Update the docblock for the constructor of the FileFailedJobProvider class. (#52149)

kevinb1989 committed:
Update the docblock for the explode method of the Stringable class. (#52148)

TENIOS committed:
Refactor PHPStan configurations (#52145)

* Refactor PHPStan configurations

* Fix quotation marks

oliverquynh committed:
Fix docblock for RoutingServiceProvider (#52144)

taylorotwell committed:
Update version to v11.16.0

taylorotwell committed:
Update CHANGELOG

siarheipashkevich committed:
Update WipeCommand.php (#52152)

https://github.com/laravel/framework/blob/11.x/src/Illuminate/Database/Console/DbCommand.php#L44

TENIOS committed:
Update outdated config files (#52150)

cosmastech committed:
[11.x] Fix 'pushProcessor method not found on LoggerInterface' error (#52117)

* add Logger@pushProcessor()

* revert IDE style

* add check to LogManager instead

* move to closure

* remove unused import

* add test

* readability

* style

siarheipashkevich committed:
Update FreshCommand.php (#52153)

https://github.com/laravel/framework/blob/11.x/src/Illuminate/Database/Console/DbCommand.php#L44

taylorotwell committed:
support boolean etag

spawnia committed:
Improve accuracy of `Collection::isEmpty` and `isNotEmpty` assertions (#52184)

* Improve accuracy of `Collection::isEmpty` and `isNotEmpty`

Those methods do not necessarily prove that the result of `first()` is not null.
However, they can prove that the result of `first()` is of the type contained within the `Collection`.

* Ensure only `first()` without parameters is affected

oliverquynh committed:
Fix return for ApplicationBuilder::withBindings method (#52181)

fernandokbs committed:
Refactor: Replace get_called_class() with static::class for consistency (#52173)

bakerkretzmar committed:
[11.x] Improve readability of SQLite schema dumps (#52172)

* Fix schema test to run against real database file

* Indent schema dump

* Wip

* Simplify test

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

* Update SchemaStateTest.php

* Update SchemaStateTest.php

---------

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>
Co-authored-by: Mior Muhammad Zaki <crynobone@gmail.com>

innocenzi committed:
feat: allow non-contextual attributes to have an `after` callback (#52167)

kingsven committed:
[11.x] Ignoring column definitions when determining if a blueprint has a create command (#52177)

* Added a test to ensure that a database column named 'create' can be added to an existing table

* Ignoring column definitions when determining if a blueprint has a create command

* Changes from StyleCI

* Update tests/Integration/Database/DatabaseSchemaBlueprintTest.php

Co-authored-by: Hafez Divandari <hafezdivandari@gmail.com>

* Update Blueprint.php

---------

Co-authored-by: Simon Geoghegan <simon@workhouseadvertising.com.au>
Co-authored-by: Taylor Otwell <taylor@laravel.com>
Co-authored-by: Hafez Divandari <hafezdivandari@gmail.com>

samuelwei committed:
Add specify exceptions for exceptions handling the vite manifest file (#52169)

* Add specify exceptions for exceptions handling the manifest files

* Fix CS

* Change to single ViteException

taylorotwell commented:
What about multi-byte strings? I would almost just make this part of `Str::limit` and rely on named arguments to enable it...

`Str::limit($value, 100, completeWords: true)`

calebdw committed:
[11.x] fix: Model newCollection generics; feat: add HasCollection trait (#52171)

* fix: newCollection generics

* feat: add HasCollection trait

taylorotwell committed:
wip

itsmewes commented:
> What about multi-byte strings? I would almost just make this part of `Str::limit` and rely on named arguments to enable it...
> 
> `Str::limit($value, 100, completeWords: true)`

I have a comment that could work with multi-byte strings. I think having this as a named argument in `Str::limit` would be the best use case. I'll update the PR to take these into account. Thanks for the response!

einar-hansen committed:
Add whereLike clause to query builder (#52147)

* Add database grammer for whereLike method

* Add whereLike to database query builder

* Add whereLike clause to MySQL grammar

* Add whereLike clause to Postgres grammar

* Add whereLike clause to SQLite grammar

* Add whereLike query builder test for sql server

* Run whereLike integration test on all databases

* Skip sqlsrv where like integration test

* Fic styling

* formatting

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

Karem-sobhy committed:
[11.x] Implement HasV7Uuids to use with MariaDB native uuid data type (#52029)

* implement HasV7UUids

* Implement all uuid4 methods to uuid7 for unified experience and add ability to define time for the uuid7

* formatting

* formatting

* remove property

---------

Co-authored-by: Taylor Otwell <taylor@laravel.com>

GromNaN committed:
Rename Model::$collection to $collectionClass (#52186)

wesleymartincc committed:
Merge branch 'str/respect-word-limit' of github-personal:itsmewes/framework into str/respect-word-limit

wesleymartincc committed:
Move logic inside limit method, use mb_ builtins

itsmewes commented:
> What about multi-byte strings? I would almost just make this part of `Str::limit` and rely on named arguments to enable it...
> 
> `Str::limit($value, 100, completeWords: true)`

This has been updated 😊

driesvints commented:
Something went wrong with branching here it seems.
